tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -64,6 +66,40 @@ private[scalajs] final class ScalaJSClassEmitter(
     linkedClassByName(className).kind == ClassKind.Interface
   }
 
+  private[emitter] lazy val candidateForJSConstructorOpt: Set[String] = {
+    /* Those classes appear in hard coded js files.
+     * Their constructors aren't always present in the desugared code.
+     * As such we cannot optimize those as we can't change the hard coded files.
+     */
+    val blackList = Set("jl_ClassCastException",
+      "sjsr_UndefinedBehaviorError", "jl_NullPointerException",
+      "js_CloneNotSupportedException")
+
+    val instantiatedClasses = linkingUnit.classDefs.filter(_.hasInstances)
+    val cantCtorOpt =
+      (instantiatedClasses.flatMap(_.superClass).map(_.name).toSet ++ blackList)
+
+    (for {
+      classDef <- linkingUnit.classDefs
```

Comment:
Although I think it does not make a different in correctness in this case (due to the predicate on constructor count), it would make more sense to filter the classDefs early with `.filter(_.kind.isClass)` to keep only Scala classes (throw out JS classes as well as interfaces).

This filter also makes sense when computing `instantiatedClasses`, so it could be done once before.